### ISSUE TITLE
Amount of simplification/normalization of goals

### ISSUE THREAD

**Issue by asr** [agda/agda#1851]
Reported by John Leon [here](http://thread.gmane.org/gmane.comp.lang.agda/8379):

``` agda
open import Data.Nat
open import Relation.Binary.PropositionalEquality

n+0=n : (n : ℕ) → n + 0 ≡ n
n+0=n zero    = refl
n+0=n (suc n) = cong suc (n+0=n (n))

m+Sn : (m n : ℕ) → m + suc n ≡ suc (m + n)
m+Sn zero    n = refl
m+Sn (suc m) n = cong suc (m+Sn m n)

+comm : (m n : ℕ) → m + n ≡ n + m
+comm zero    n =
  let a = sym (n+0=n n)
  in {!!}
+comm (suc m) n =
  let a = cong suc (+comm m n)
      b = sym (m+Sn n m)
  in {!!}
```

Using `C-c C-,` in the first goal, master and 2.4.2.5 have different behaviours. In master the type of `a` is not normalised

``` agda
Goal: n ≡ n + zero
————————————————————————————————————————————————————————————
a : .Function.flip _≡_ (n + 0) n
n : ℕ
```

but in 2.4.2.5 the type of `a` is full normalised

``` agda
Goal: n ≡ n + zero
————————————————————————————————————————————————————————————
a : n ≡ n + 0
n : ℕ
```

From the CHANGELOG for 2.4.0

```
* Key bindings for controlling simplification/normalisation:

  [TODO: Simplification should be explained somewhere.]

  Commands like "Goal type and context" (C-c C-,) could previously be
  invoked in two ways. By default the output was normalised, but if a
  prefix argument was used (for instance via C-u C-c C-,), then no
  explicit normalisation was performed. Now there are three options:

  * By default (C-c C-,) the output is simplified.

  * If C-u is used exactly once (C-u C-c C-,), then the result is
    neither (explicitly) normalised nor simplified.

  * If C-u is used twice (C-u C-u C-c C-,), then the result is
    normalised.
```

the type of `a` should be not normalised in both master and 2.4.2.5.

Why the difference between master and 2.4.2.5? Is the CHANGELOG for 2.4.0 correct? Is there a test case for the changes in 2.4.0?

**Comment by nad** [comment]
> Is there a test case for the changes in 2.4.0?

AFAIK the Emacs Lisp code is not tested (by `make test`).

I found one test of simplification, see `test/interaction/Long.in`.

**Comment by nad** [comment]
By the way, I would assume that this has nothing to do with the Emacs Lisp code.

**Comment by UlfNorell** [comment]
What is the issue here? Is it deciding what the key bindings should be for 2.5.1? Right now in master:
- `(no modifier)`: simplification
- `C-u`: non-normalised
- `C-u C-u`: normalised

I don't have strong opinions on this, but personally I use normalised more than the other two.

**Comment by asr** [comment]
> What is the issue here?

Why master and 2.4.2.5 have different behaviours if the change was introduced in 2.4.0.

**Comment by asr** [comment]
> Right now in master:
> - `(no modifier)`: simplification
> - `C-u`: non-normalised
> - `C-u C-u`: normalised
> 
> I don't have strong opinions on this, but personally I use normalised more than the ther two.

I filed #1868 for discussing about the key bindings.

**Comment by UlfNorell** [comment]
Is this important? Isn't the behaviour of master the intended one?

**Comment by asr** [comment]
> Is this important?

At least we should know why the behaviour changed and to add a test case if possible.

> Isn't the behaviour of master the intended one?

Yes, it is.

**Comment by nad** [comment]
I seem to get output like `const <something> <something else>` quite often now. I think the reason is often that `_×_` is defined in the following way: `A × B = Σ A (const B)`. Are you saying that this is intended?

**Comment by UlfNorell** [comment]
I don't know what simplification is doing and what its intended behaviour is. It wasn't clear to me from the OP that this issue is about what the simplifier is doing.

**Comment by asr** [comment]
What did you mean by 'OP'?

**Comment by UlfNorell** [comment]
OP is short for "original post" (or poster) and is commonly used in forums to refer to the first post in a thread.

**Comment by asr** [comment]
I'll write the test case...

**Comment by UlfNorell** [comment]
I get the same behaviour in both master and `2.4.2.5`. Here's the stand-alone example I'm using:

``` agda
postulate
  Nat : Set
  _+_ : Nat → Nat → Nat

data _≡_ {a} {A : Set a} (x : A) : A → Set where
  refl : x ≡ x

flip : ∀ {a b c} {A : Set a} {B : Set b} {C : A → B → Set c} →
       ((x : A) (y : B) → C x y) → ((y : B) (x : A) → C x y)
flip f = λ y x → f x y

+comm : (n : Nat) → flip _≡_ (n + n) n → Set
+comm n eq = {!!}  -- eq : flip _≡_ (n + n) n
```

Closing this tentatively. Please reopen if you have a self-contained test case where master and 2.4.2.5 differ.

**Comment by asr** [comment]
> Please reopen if you have a self-contained test case where master and 2.4.2.5 differ.

This is a self-contained test case.

``` agda
-- Adapted from an example reported by John Leon

open import Agda.Primitive

infix 4 _≡_
data _≡_ {a} {A : Set a} (x : A) : A → Set a where
  refl : x ≡ x

cong : ∀ {a b} {A : Set a} {B : Set b}
       (f : A → B) {x y} → x ≡ y → f x ≡ f y
cong f refl = refl

REL : ∀ {a b} → Set a → Set b → (ℓ : Level) → Set (a ⊔ b ⊔ lsuc ℓ)
REL A B ℓ = A → B → Set ℓ

Rel : ∀ {a} → Set a → (ℓ : Level) → Set (a ⊔ lsuc ℓ)
Rel A ℓ = REL A A ℓ

flip : ∀ {a b c} {A : Set a} {B : Set b} {C : A → B → Set c} →
       ((x : A) (y : B) → C x y) → ((y : B) (x : A) → C x y)
flip f = λ y x → f x y

infixr 4 _⇒_
_⇒_ : ∀ {a b ℓ₁ ℓ₂} {A : Set a} {B : Set b} →
      REL A B ℓ₁ → REL A B ℓ₂ → Set _
P ⇒ Q = ∀ {i j} → P i j → Q i j

Sym : ∀ {a b ℓ₁ ℓ₂} {A : Set a} {B : Set b} →
      REL A B ℓ₁ → REL B A ℓ₂ → Set _
Sym P Q = P ⇒ flip Q

Symmetric : ∀ {a ℓ} {A : Set a} → Rel A ℓ → Set _
Symmetric _∼_ = Sym _∼_ _∼_

sym : ∀ {a} {A : Set a} → Symmetric (_≡_ {A = A})
sym refl = refl

data ℕ : Set where
  zero : ℕ
  suc  : (n : ℕ) → ℕ

infixl 6 _+_
_+_ : ℕ → ℕ → ℕ
zero  + m = m
suc n + m = suc (n + m)

n+0=n : (n : ℕ) → n + zero ≡ n
n+0=n zero    = refl
n+0=n (suc n) = cong suc (n+0=n (n))

m+Sn : (m n : ℕ) → m + suc n ≡ suc (m + n)
m+Sn zero    n = refl
m+Sn (suc m) n = cong suc (m+Sn m n)

+comm : (m n : ℕ) → m + n ≡ n + m
+comm zero    n =
  let a = sym (n+0=n n)
  in {!!}
+comm (suc m) n =
  let a = cong suc (+comm m n)
      b = sym (m+Sn n m)
  in {!!}

-- Using `C-c C-,` in the first goal, master and 2.4.2.5 have different
-- behaviours. In master, the type of `a` is not normalised

-- Goal: n ≡ n + zero
-- ————————————————————————————————————————————————————————————
-- a : flip _≡_ (n + zero) n
-- n : ℕ

-- but in 2.4.2.5, the type of `a` is full normalised

-- Goal: n ≡ n + zero
-- ————————————————————————————————————————————————————————————
-- a : n ≡ n + 0
-- n : ℕ
```

Let's see what `git bisect` says...

**Comment by asr** [comment]
Added `infixl 6 _+_` to the above example because it was required when running `git bisect`.

**Comment by asr** [comment]
A simplified test-case by removing the dependency on `Agda.Primitive`:

``` agda
-- Adapted from an example reported by John Leon

infix 4 _≡_
data _≡_ {A : Set} (x : A) : A → Set where
  refl : x ≡ x

cong : ∀ {A B : Set} (f : A → B) {x y} → x ≡ y → f x ≡ f y
cong f refl = refl

REL : Set → Set → Set₁
REL A B = A → B → Set

flip : {A B : Set} {C : A → B → Set₁} →
       ((x : A) (y : B) → C x y) → ((y : B) (x : A) → C x y)
flip f = λ y x → f x y

infixr 4 _⇒_
_⇒_ : {A B : Set} → REL A B → REL A B → Set
P ⇒ Q = ∀ {i j} → P i j → Q i j

Sym : {A B : Set} → REL A B → REL B A → Set
Sym P Q = P ⇒ flip Q

Symmetric : {A : Set} → REL A A → Set
Symmetric _∼_ = Sym _∼_ _∼_

sym : {A : Set} → Symmetric (_≡_ {A = A})
sym refl = refl

data ℕ : Set where
  zero : ℕ
  suc  : (n : ℕ) → ℕ

infixl 6 _+_
_+_ : ℕ → ℕ → ℕ
zero  + m = m
suc n + m = suc (n + m)

n+0=n : (n : ℕ) → n + zero ≡ n
n+0=n zero    = refl
n+0=n (suc n) = cong suc (n+0=n (n))

m+Sn : (m n : ℕ) → m + suc n ≡ suc (m + n)
m+Sn zero    n = refl
m+Sn (suc m) n = cong suc (m+Sn m n)

+comm : (m n : ℕ) → m + n ≡ n + m
+comm zero    n =
  let a = sym (n+0=n n)
  in {!!}
+comm (suc m) n =
  let a = cong suc (+comm m n)
      b = sym (m+Sn n m)
  in {!!}

-- Using `C-c C-,` in the first goal, master and 2.4.2.5 have different
-- behaviours. In master, the type of `a` is not normalised

-- Goal: n ≡ n + zero
-- ————————————————————————————————————————————————————————————
-- a : flip _≡_ (n + zero) n
-- n : ℕ

-- but in 2.4.2.5, the type of `a` is full normalised

-- Goal: n ≡ n + zero
-- ————————————————————————————————————————————————————————————
-- a : n ≡ n + 0
-- n : ℕ
```

**Comment by asr** [comment]
I added a [test case](https://github.com/agda/agda/commit/d76b3920f20624ee630f1a90ad69ce97d2a386db).

It seems that change was caused by 344296c.

For the record: I couldn't find the "bad commit" using `git bisect` because from the OP reported [here](http://thread.gmane.org/gmane.comp.lang.agda/8379) the change occurred between 2.5.0.20160213 (f8ee850) and 2.4.2.5 (75cb744), but the "bad commit" reported above (i.e. 344296c) is not within that range.

Since no one is dissatisfied with the current behaviour, I'm closing the issue.
